MAINT: Review F401,F841,F842 flake8 errors (unused variables and imports)

rth opened the pull request:
In this PR I run the flake8 on checks corresponding to unused imports and variables,
```
flake8 --select=F401,F841,F842 numpy
```
and selectively removed those I thought should be harmless. For unused variables that looked suspicious, a `FIXME` comment was added.

Unused imports is an issue beyond code style as it makes dependencies between modules harder to understand (not that there much issues with that in numpy -- this PR mostly changes test files).


This should also help with LGTM.com warnings.

rth committed:
Review F401,F841,F842 flake8 errors (unused variables, imports)

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4868,7 +4867,7 @@ def test_field_rename(self):
     def test_multiple_field_name_occurrence(self):
         def test_assign():
-            dtype = np.dtype([("A", "f8"), ("B", "f8"), ("A", "f8")])
+            np.dtype([("A", "f8"), ("B", "f8"), ("A", "f8")])
```
Test name should be changed here any below to not include the word `assign`, since it no longer does an assignment

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -7517,7 +7514,7 @@ def test_argmin_with_out(self):
         out = np.eye(5, dtype='i2')
-        res = np.clip(mat, a_min=-10, a_max=0, out=out)
+        np.clip(mat, a_min=-10, a_max=0, out=out)
         assert_equal(np.sum(out), 0)
```
Would be better to add `assert_(res is out)` here instead

eric-wieser commented on numpy/core/tests/test_scalarinherit.py:
```diff
@@ -67,7 +67,6 @@ class Mystr(str, np.generic):
 
     def test_char_repeat(self):
         np_s = np.string_('abc')
-        np_u = np.unicode_('abc')
```
Instead of removing this line, it looks like the intent was to test it below.

eric-wieser commented on numpy/core/tests/test_scalarmath.py:
```diff
@@ -184,7 +184,6 @@ def test_modular_power(self):
         a = 5
         b = 4
         c = 10
-        expected = pow(a, b, c)
```
Would be inclined to leave this as documentation. Feel free to stick whatever the comment is that flake8 needs to ignore this line.

eric-wieser commented on numpy/distutils/misc_util.py:
```diff
@@ -2225,7 +2224,6 @@ def is_bootstrapping():
         return True
     except AttributeError:
         return False
-        __NUMPY_SETUP__ = False
```
Pretty sure this one is important, and used by another file.

charris replied:
It is local to the function and can have no effect. Looks like a cut and paste refactor bug.

eric-wieser commented on numpy/distutils/system_info.py:
```diff
@@ -1125,7 +1125,6 @@ def calc_info(self):
         atlas_1 = None
         for d in lib_dirs:
             atlas = self.check_libs2(d, atlas_libs, [])
-            lapack_atlas = self.check_libs2(d, ['lapack_atlas'], [])
```
The fact this line is unused looks like a bug to me

eric-wieser commented on numpy/distutils/tests/test_misc_util.py:
```diff
@@ -78,4 +78,4 @@ def test_get_shared_lib_extension(self):
     # Regression test for gh-7707.  If npymath.ini wasn't installed, then this
     # will give an error.
-    info = get_info('npymath')
+    get_info('npymath')
```
Would be great to add some basic test of `info` here, rather than just ignoring the return value

eric-wieser commented:
Mostly looks good. Thanks for only touching test and private files - note that in many cases we're stuck with unused imports in our public files because those import often accidentally form part of our public API.

rth committed:
Review comments

rth committed:
More tests in test_installed_npymath_ini

rth commented:
Thanks for the review! I addressed your comments, added a few `FIXME` tags for things that you marked as potentially suspicious and that I didn't want to investigate in this PR.

> note that in many cases we're stuck with unused imports in our public files because those import often accidentally form part of our public API.

Yes, it's not an easy problem...

rth committed:
Merge branch 'master' into dead-code

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4932,9 +4931,7 @@ def test_bytes_fields(self):
     @pytest.mark.skipif(sys.version_info[0] < 3, reason="Not Python 3")
     def test_multiple_field_name_unicode(self):
         def test_assign_unicode():
```
Similarly here - there's no assignment any more

charris replied:
@rth Still needs a fix.

charris replied:
NVM, looks fixed.

rth replied:
I did fix this is the last commit I think.

eric-wieser commented on numpy/core/tests/test_scalarinherit.py:
```diff
@@ -69,6 +69,7 @@ def test_char_repeat(self):
-        res_np = np_s * np_i
         res_s = b'abc' * 5
-        assert_(res_np == res_s)
+        res_u = 'abc' * 5
```
Needs a `u` prefix to test the right thing on python 2

eric-wieser commented on numpy/distutils/misc_util.py:
```diff
@@ -2225,6 +2224,7 @@ def is_bootstrapping():
     except AttributeError:
         return False
+        # FIXME: below code is never reached
         __NUMPY_SETUP__ = False
```
Ok, I get this now - I think this is supposed to set `builtins.__NUMPY_SETUP__`.

charris replied:
Let's leave it out, it hasn't been missed since 2009.

charris commented:
@rth Ping.

rth committed:
Review comments

rth commented:
Thanks for the review @eric-wieser and @charris !
